File: 字节跳动高频/电话号码的字母组合.py
class Solution:
    def __init__(self):
        self.num2str = ["abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"]

    def combine(self, res: List[int], cur: str, digits: str, start: int):
        if start == len(digits) - 1 :
            for c in self.num2str[int(digits[start]) - 2]:
                res.append(cur + c)
        else:
            for c in self.num2str[int(digits[start]) - 2]:
                self.combine(res, cur + c, digits, start + 1)

    def letterCombinations(self, digits: str) -> List[str]:
        if len(digits) == 0:
            return []
        res = []
        cur = ''
        self.combine(res, cur, digits, 0)
        return res

File: 字节跳动高频/电话号码的字母组合.cpp
class Solution {
    string num2str[8] = {"abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    void combine(vector<string>& res, string cur, string digits, int start) {
        if (start == digits.length() - 1) {
            for (char c : this->num2str[digits[start] - '0' - 2]) {
                res.emplace_back(cur + c);
            }
        }
        else {
            for (char c : this->num2str[digits[start] - '0' - 2]) {
                this->combine(res, cur + c, digits, start + 1);
            }
        }

    }
public:
    vector<string> letterCombinations(string digits) {
        vector<string> res;
        if (digits.length() == 0) return res;
        string cur = "";
        this->combine(res, cur, digits, 0);
        return res;
    }
};